File: res/base/scripts/screens/main_screen.lua
require("containers/system_tools")

class "MainScreen"(Screen)

MainScreen.desc = 
{
    ui = {
        {
            type = "Button",
            id = "testBtn",
            text = "Test Screen",
            font = "system_15_fnt",
            rect = {0, 0, 256, 64},
            colour = "green",
            text_align = {"CENTER|MIDDLE"},
            sprites = {"up_btn_spr", "down_btn_spr", "over_btn_sptr"},
            callback = {"MouseUp_Left", "MainScreen.onTestBtnClicked"}
        },
        {
            type = "Button",
            id = "newGameBtn",
            text = "New Game",
            font = "system_15_fnt",
            rect = {0, 0, 256, 64},
            align = {"RIGHT|TOP", -64, 64},
            colour = "white",
            text_align = {"CENTER|MIDDLE"},
            sprites = {"up_btn_spr", "down_btn_spr", "over_btn_sptr"},
            callback = {"MouseUp_Left", "MainScreen.onNewGameBtnClicked"}
        },
        {
            type = "Button",
            id = "loadGameBtn",
            text = "Load Game",
            font = "system_15_fnt",
            rect = {0, 0, 256, 64},
            align = {"RIGHT|TOP", -64, 138},
            colour = "white",
            text_align = {"CENTER|MIDDLE"},
            sprites = {"up_btn_spr", "down_btn_spr", "over_btn_sptr"},
            callback = {"MouseUp_Left", "MainScreen.onLoadGameBtnClicked"}
        },
        {
            type = "Button",
            id = "optionsBtn",
            text = "Options",
            font = "system_15_fnt",
            rect = {0, 0, 256, 64},
            align = {"RIGHT|TOP", -64, 212},
            colour = "white",
            text_align = {"CENTER|MIDDLE"},
            sprites = {"up_btn_spr", "down_btn_spr", "over_btn_sptr"},
            callback = {"MouseUp_Left", "MainScreen.onOptionsBtnClicked"}
        },
        {
            type = "Button",
            id = "mapEditorBtn",
            text = "Map Editor",
            font = "system_15_fnt",
            rect = {0, 0, 256, 64},
            align = {"RIGHT|TOP", -64, 286},
            colour = "white",
            text_align = {"CENTER|MIDDLE"},
            sprites = {"up_btn_spr", "down_btn_spr", "over_btn_sptr"},
            callback = {"MouseUp_Left", "MainScreen.onMapEditorBtnClicked"}
        },
        {
            type = "Button",
            id = "exitBtn",
            text = "Exit",
            font = "system_15_fnt",
            rect = {0, 0, 256, 64},
            align = {"RIGHT|TOP", -64, -64},
            colour = "red",
            text_align = {"CENTER|MIDDLE"},
            sprites = {"up_btn_spr", "down_btn_spr", "over_btn_sptr"},
            callback = {"MouseUp_Left", "MainScreen.onExitBtnClicked"}
        }
        }
}

function MainScreen:init()
    local btn = Button("testBtn")
    btn:setText("Test Screen")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setColour("green")
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onTestBtnClicked, self)
    self:attach(btn)

    btn = Button("newGameBtn")
    btn:setText("New Game")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setAlignment("RIGHT|TOP", -64, 64)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onNewGameBtnClicked, self)
    self:attach(btn)

    btn = Button("loadGameBtn")
    btn:setText("Load Game")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setAlignment("RIGHT|TOP", -64, 138)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onLoadGameBtnClicked, self)
    self:attach(btn)

    btn = Button("optionsBtn")
    btn:setText("Options")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setAlignment("RIGHT|TOP", -64, 212)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onOptionsBtnClicked, self)
    self:attach(btn)

    btn = Button("mapEditorBtn")
    btn:setText("Map Editor")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setAlignment("RIGHT|TOP", -64, 286)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onMapEditorBtnClicked, self)
    self:attach(btn)

    btn = Button("exitBtn")
    btn:setText("Exit")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setColour("red")
    btn:setAlignment("RIGHT|BOTTOM", -64, -64)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onExitBtnClicked, self)
    self:attach(btn)
end

function MainScreen:onExitBtnClicked()
    Engine.addCommand({id = "shutdown"})
end

function MainScreen:onMapEditorBtnClicked()
    Screens.load("LoadingScreen", "MapEditorScreen")
end

function MainScreen:onOptionsBtnClicked()
    Screens.load("LoadingScreen", "OptionsScreen")
end

function MainScreen:onLoadGameBtnClicked()
end

function MainScreen:onNewGameBtnClicked()
end

function MainScreen:onTestBtnClicked()
    Screens.load("LoadingScreen", "TestScreen")
end

File: src/main_screen.cpp
#include "main_screen.h"

#include "system_tools.h"

namespace redrevolt
{
    MainScreen::MainScreen(const std::string & id)
        : stren::Screen(id)
    {
        SystemTools * systemTools = new SystemTools("systemTools");
        attach(systemTools);
        {
            stren::Button * btn = new stren::Button("testBtn");
            btn->setText("Test Screen");
            btn->setTextAlignment("CENTER|MIDDLE");
            btn->setFont("system_15_fnt");
            btn->setRect(0, 0, 256, 64);
            btn->setAlignment("LEFT|TOP", 64, 64);
            btn->setColour("green");
            btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
            btn->addCallback("MouseUp_Left", this, &MainScreen::onTestBtnClicked);
            attach(btn);
        }
        {
            stren::Button * btn = new stren::Button("newGameBtn");
            btn->setText("New Game");
            btn->setTextAlignment("CENTER|MIDDLE");
            btn->setFont("system_15_fnt");
            btn->setRect(0, 0, 256, 64);
            btn->setAlignment("RIGHT|TOP", -64, 64);
            btn->setColour("white");
            btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
            btn->addCallback("MouseUp_Left", this, &MainScreen::onNewGameBtnClicked);
            attach(btn);
        }
        {
            stren::Button * btn = new stren::Button("loadGameBtn");
            btn->setText("Load Game");
            btn->setTextAlignment("CENTER|MIDDLE");
            btn->setFont("system_15_fnt");
            btn->setRect(0, 0, 256, 64);
            btn->setAlignment("RIGHT|TOP", -64, 138);
            btn->setColour("white");
            btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
            btn->addCallback("MouseUp_Left", this, &MainScreen::onLoadGameBtnClicked);
            attach(btn);
        }
        {
            stren::Button * btn = new stren::Button("optionsBtn");
            btn->setText("Options");
            btn->setTextAlignment("CENTER|MIDDLE");
            btn->setFont("system_15_fnt");
            btn->setRect(0, 0, 256, 64);
            btn->setAlignment("RIGHT|TOP", -64, 212);
            btn->setColour("white");
            btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
            btn->addCallback("MouseUp_Left", this, &MainScreen::onOptionsBtnClicked);
            attach(btn);
        }
        {
            stren::Button * btn = new stren::Button("mapEditorBtn");
            btn->setText("Map Editor");
            btn->setTextAlignment("CENTER|MIDDLE");
            btn->setFont("system_15_fnt");
            btn->setRect(0, 0, 256, 64);
            btn->setAlignment("RIGHT|TOP", -64, 286);
            btn->setColour("white");
            btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
            btn->addCallback("MouseUp_Left", this, &MainScreen::onMapEditorBtnClicked);
            attach(btn);
        }
        {
            stren::Button * btn = new stren::Button("exitBtn");
            btn->setText("Exit");
            btn->setTextAlignment("CENTER|MIDDLE");
            btn->setFont("system_15_fnt");
            btn->setRect(0, 0, 256, 64);
            btn->setAlignment("RIGHT|BOTTOM", -64, -64);
            btn->setColour("red");
            btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
            btn->addCallback("MouseUp_Left", this, &MainScreen::onExitBtnClicked);
            attach(btn);
        }
    }

    MainScreen::~MainScreen()
    {
    }

    void MainScreen::onTestBtnClicked(Widget * sender)
    {
        stren::SwitchScreenCommand command;
        command.setScreen("LoadingScreen");
        command.setNextScreen("TestScreen");
        command.execute();
    }

    void MainScreen::onNewGameBtnClicked(Widget * sender)
    {
        log("new game");
    }

    void MainScreen::onLoadGameBtnClicked(Widget * sender)
    {
        log("load game");
    }

    void MainScreen::onOptionsBtnClicked(Widget * sender)
    {
        stren::SwitchScreenCommand command;
        command.setScreen("LoadingScreen");
        command.setNextScreen("OptionsScreen");
        command.execute();
    }

    void MainScreen::onMapEditorBtnClicked(Widget * sender)
    {
        stren::SwitchScreenCommand command;
        command.setScreen("LoadingScreen");
        command.setNextScreen("MapEditorScreen");
        command.execute();
    }

    void MainScreen::onExitBtnClicked(Widget * sender)
    {
        stren::Engine::shutdown();
    }

    void MainScreen::log(const char * msg)
    {
        stren::Widget * widget = findWidget("systemTools");
        if (widget)
        {
            SystemTools * tools = dynamic_cast<SystemTools *>(widget);
            if (tools)
            {
                tools->log(msg);
            }
        }
    }
} // redrevolt

